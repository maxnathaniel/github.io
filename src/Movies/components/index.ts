export * from './Movies';
export * from './MovieItem';
export * from './MovieDetails';