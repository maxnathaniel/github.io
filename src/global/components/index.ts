export * from './App';
export * from './PlaceholderRow';