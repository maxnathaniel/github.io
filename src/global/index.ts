export * from './api';
export * from './styles';
export * from './components';
export * from './utils';