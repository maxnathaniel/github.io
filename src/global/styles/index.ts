export * from './base-styles';
export * from './global-styles';