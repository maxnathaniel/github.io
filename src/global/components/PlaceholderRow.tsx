import React, { FC} from 'react';
import { Placeholder } from 'semantic-ui-react';

import { styled } from '../styles';
import { device } from '../utils';

const PlaceholdeContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

const PlaceholderItemWrapper = styled.div`
  min-height: 300px;
  margin: 10px;
  width: calc( (100% / 4) - 20px );
  padding: 10px;
  cursor: pointer;
  transition: all 0.2s ease-in;

  @media only screen and ${device.mobileM} {
    width: calc(100% - 20px);
  }

  @media only screen and ${device.mobileL} {
    width: calc( (100% / 2) - 20px );
  }

  @media only screen and ${device.laptop} {
    width: calc( (100% / 3) - 20px );
  }
`;

export const PlaceholderRow: FC = () => {
  return (
    <PlaceholdeContainer>
      <PlaceholderItemWrapper>
        <Placeholder inverted style={{ width: '100%', height: '100%'}}>
          <Placeholder.Image />
        </Placeholder>
      </PlaceholderItemWrapper>
      <PlaceholderItemWrapper>
        <Placeholder inverted style={{ width: '100%', height: '100%'}}>
          <Placeholder.Image />
        </Placeholder>
        </PlaceholderItemWrapper>
      <PlaceholderItemWrapper>
        <Placeholder inverted style={{ width: '100%', height: '100%'}}>
          <Placeholder.Image />
        </Placeholder>
      </PlaceholderItemWrapper>
    </PlaceholdeContainer>
  );
};

export default PlaceholderRow;