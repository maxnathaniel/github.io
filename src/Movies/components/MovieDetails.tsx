import React, { FC, useEffect, useState } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { Breadcrumb, Image, Placeholder } from 'semantic-ui-react';

import { device, getMovieDetails, styled } from 'src/global';
import { SearchResults } from './Movies';

const MovieDetailsContainer = styled.div`
  width: 100%;
  height: 100%;
  text-align: center;
`;

const Section = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: 80%;
  margin: 2% auto;
  align-items: center;
  justify-content: center;
`;

const ErrorText = styled.p`
  margin: 20% auto 0 auto;
  font-size: 1.6em;
  color: ${(props) => props.theme.colours.red1};
`;

const Poster = styled.div`
  flex: 1 1 auto;
  min-width: 40%;
  max-width: 40%;
`;

const HeaderText = styled.h1`
  font-family: Open Sans;
  font-size: 2em;
  text-align: left;
  max-width: 550px;

  @media only screen and ${device.mobileM} {
    text-align: center;
    max-width: 100%;
  }

  @media only screen and ${device.mobileL} {
    text-align: center;
    max-width: 100%;
  }

  @media only screen and ${device.tablet} {
    text-align: center;
    max-width: 400px;
  }
`;

const SecondayHeaderText = styled.h1`
  font-family: Open Sans;
  font-size: 1.5em;
  font-weight: 300;
`;

const PosterContent = styled.div`
  flex: 2 1 auto;
  align-self: flex-start;
`;

const MoviePlot = styled.div`
  flex: 1 1 auto;
  min-width: 40%;
  max-width: 40%;
  padding: 10px;
  align-self: flex-start;
  text-align: left;

  @media only screen and ${device.mobileM} {
    min-width: 100%;
  }

  @media only screen and ${device.mobileL} {
    min-width: 100%;

    > p {
      max-width: 100%;
    }
  }

  @media only screen and ${device.tablet} {
    width: 100%;

    > p {
      max-width: 100%;
    }
  }

  @media only screen and ${device.laptop} {
    min-width: 40%;
    max-width: 40%;

    > p {
      max-width: 350px;
    }
  }
`;

const Border = styled.div`
  width: 80%;
  margin: 0 auto;
  border: 1px solid ${(props) => props.theme.colours.grey1};
`;

const MovieDetailsText = styled.div`
  flex: 2 1 auto;
  padding: 10px;
  margin-top: 20px;
`;

const Label = styled.p`
  flex: 1 1 auto;
  color: ${(props) => props.theme.colours.neon1};
  text-align: left;
  max-width: 100px;
`;

const Text = styled.p`
  flex: 2 1 auto;
  text-align: left;
  max-width: 350px;

  @media only screen and ${device.mobileM} {
    max-width: 250px;
  }

  @media only screen and ${device.mobileL} {
    max-width: 200px;
  }
`;

const FlexRow = styled.div`
  display: flex;
  flex-direction: row;
  margin: 5px 0;
`;

const Source = styled.h4`
  display: inline-block;
`;

const RatingsWrapper = styled.div`
  text-align: left;
  margin: 10px 10px 0 0;
`;

const Circle = styled.div`
  display: inline-block;
  border-radius: 50%;
  height: 60px;
  width: 60px;
  background-color: ${(props) => props.theme.colours.neon1};
  vertical-align: middle;
  text-align: center;
  margin-right: 20px;
  color: black;

  > p {
    margin-top: 30%;
  }
`;

const PulsingCircle = styled.div`
  display: inline-block;
  border-radius: 50%;
  height: 60px;
  width: 60px;
  background-color: ${(props) => props.theme.colours.neon1};
  vertical-align: middle;
  text-align: center;
  margin-right: 20px;
  color: black;
  box-shadow: 0 0 0 0 rgba(0, 0, 0, 1);
  transform: scale(1);
  animation: pulse 2s infinite;

  @keyframes pulse {
    0% {
      transform: scale(0.85);
      box-shadow: 0 0 0 0 rgba(0, 0, 0, 0.7);
    }

    70% {
      transform: scale(1);
      box-shadow: 0 0 0 20px rgba(0, 0, 0, 0);
    }

    100% {
      transform: scale(0.85);
      box-shadow: 0 0 0 0 rgba(0, 0, 0, 0);
    }
  }

  > p {
    margin-top: 30%;
  }
`;

const ActorWrapper = styled.div`
  flex: 1 1 auto;
  text-align: left;
  margin: 10px 0;
`;

interface MovieDetailsData extends SearchResults {
  Rated: string;
  Released: string;
  Runtime: string;
  Genre: string;
  Director: string;
  Writer: string;
  Actors: string;
  Plot: string;
  Language: string;
  Country: string;
  Awards: string;
  Ratings: any[];
  Metascore: string;
  imdbRating: string;
  imdbVotes: string;
  Type: string;
  DVD: string;
  BoxOffice: string;
  Production: string;
  Website: string;
  Response: string;
}

export const MovieDetails: FC = () => {
  const { id } = useParams();
  const history = useHistory();
  const [isLoading, setIsLoading] = useState(false);
  const [details, setDetails] = useState<MovieDetailsData | null>(null);
  const [error, setError] = useState('');

  const onReturn = () => {
    history.push('../');
  };

  useEffect(() => {
    try {
      setIsLoading(true);
      getMovie(id);
    } catch (e) {
      console.error('Unable to retrieve movie details', e);
      setError(e.toString());
    }
    setIsLoading(false);

    async function getMovie(id: string) {
      try {
        const data = await getMovieDetails(id);
        if (data) {
          setDetails(data);
        }
      } catch (e) {
        setError('Unable to retrieve movie details');
      }
    }
  }, []);

  return (
    <MovieDetailsContainer>
      <Breadcrumb
        style={{ marginTop: '10px', width: '80%', textAlign: 'left' }}
      >
        <Breadcrumb.Section
          link
          onClick={onReturn}
          style={{ color: '#03dac5' }}
        >
          Home
        </Breadcrumb.Section>
        <Breadcrumb.Divider style={{ color: '#03dac5' }} />
        <Breadcrumb.Section active>Movie</Breadcrumb.Section>
      </Breadcrumb>
      {error && <ErrorText>Unable to retrieve movie details</ErrorText>}
      {details && (
        <>
          <Section>
            <Poster>
              {details.Poster === 'N/A' ? (
                <Placeholder
                  inverted
                  style={{ maxWidth: '250px', minHeight: '300px' }}
                >
                  <Placeholder.Image />
                </Placeholder>
              ) : (
                <Image
                  style={{ margin: '0 auto' }}
                  src={details.Poster}
                  size='medium'
                />
              )}
            </Poster>
            <PosterContent>
              <HeaderText>{details.Title}</HeaderText>
              {details.Ratings.length &&
                details.Ratings.map((rating) => {
                  return (
                    <RatingsWrapper key={rating}>
                      <PulsingCircle>
                        <p>{rating.Value}</p>
                      </PulsingCircle>
                      <Source>
                        <p>{rating.Source}</p>
                      </Source>
                    </RatingsWrapper>
                  );
                })}
            </PosterContent>
          </Section>
          <Border />
          <Section>
            <MoviePlot>
              <SecondayHeaderText>Movie Details</SecondayHeaderText>
              <Text style={{ color: '#03dac5' }}>Synposis</Text>
              <Text style={{ color: '#FFF', lineHeight: '2' }}>
                {details.Plot}
              </Text>
            </MoviePlot>
            <MovieDetailsText>
              <SecondayHeaderText>{''}</SecondayHeaderText>
              <FlexRow>
                <Label>Awards</Label>
                <Text>{details.Awards}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Rated</Label>
                <Text>{details.Rated}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Genre</Label>
                <Text>{details.Genre}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Director</Label>
                <Text>{details.Director}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Writer</Label>
                <Text>{details.Writer}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Released</Label>
                <Text>{details.Released}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Box Office</Label>
                <Text>{details.BoxOffice}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Runtime</Label>
                <Text>{details.Runtime}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Studio</Label>
                <Text>{details.Production}</Text>
              </FlexRow>
              <FlexRow>
                <Label>Country</Label>
                <Text>{details.Country}</Text>
              </FlexRow>
            </MovieDetailsText>
          </Section>
          <Border />
          <Section>
            <SecondayHeaderText style={{ width: '100%', textAlign: 'left' }}>
              Cast
            </SecondayHeaderText>
            <FlexRow style={{ width: '100%', flexWrap: 'wrap' }}>
              {details.Actors.split(',').map((actor) => {
                return (
                  <ActorWrapper key={actor} style={{ minWidth: '25%' }}>
                    <Circle>
                      <p>{actor.trim()[0]}</p>
                    </Circle>
                    <Text style={{ display: 'inline' }}>{actor.trim()}</Text>
                  </ActorWrapper>
                );
              })}
            </FlexRow>
          </Section>
        </>
      )}
    </MovieDetailsContainer>
  );
};

export default MovieDetails;
