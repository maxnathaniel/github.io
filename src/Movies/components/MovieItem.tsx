import React, { FC, useCallback, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { Image, Placeholder } from 'semantic-ui-react';

import { styled, device } from 'src/global';
import { SearchResults } from './Movies';

const MovieItemContainer = styled.div`
  min-height: 300px;
  margin: 10px;
  padding: 10px;
  cursor: pointer;
  transition: all 0.2s ease-in;

  @media only screen and ${device.mobileM} {
    width: calc(100% - 20px);
  }

  @media only screen and ${device.mobileL} {
    width: calc((100% / 2) - 20px);
  }

  @media only screen and ${device.laptop} {
    width: calc((100% / 3) - 20px);
  }

  :hover {
    transform: scale(1.1);
  }
`;

const ImageContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 100%;
`;

interface Props {
  data?: Partial<SearchResults> | undefined;
  isPlaceHolder?: boolean;
}

export const MovieItem: FC<Props> = (props) => {
  const { data, isPlaceHolder = false } = props;
  const history = useHistory();

  const onClick = useCallback(() => {
    if (data?.imdbID) {
      history.push(`/movie/${data.imdbID}`);
    }
  }, []);

  return (
    <MovieItemContainer>
      {data && data.Poster === 'N/A' ? (
        <Placeholder
          onClick={onClick}
          inverted
          style={{ height: '100%', width: '100%' }}
        >
          <Placeholder.Image />
        </Placeholder>
      ) : (
        <ImageContainer>
          <Image
            style={{ border: '2px solid #03DAC5' }}
            src={data?.Poster}
            alt='movie-img'
            size='big'
            onClick={onClick}
          />
        </ImageContainer>
      )}
      {isPlaceHolder && (
        <Placeholder inverted style={{ height: '100%', width: '100%' }}>
          <Placeholder.Image />
        </Placeholder>
      )}
    </MovieItemContainer>
  );
};
