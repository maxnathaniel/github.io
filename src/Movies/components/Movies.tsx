import React, { FC, useEffect, useRef, useState } from 'react';

import { getMovies, styled } from 'src/global';
import { SearchBar } from 'src/Search';
import { MovieItem } from './MovieItem';

const MoviesContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

const WelcomeText = styled.p`
  font-size: 1.8em;
  color: ${(props) => props.theme.colours.neon1};
  margin: 20% auto 0 auto;
`;

const ErrorText = styled.p`
  margin: 20% auto 0 auto;
  font-size: 1.6em;
  color: ${(props) => props.theme.colours.red1};
`;

const Bottom = styled.div``;

export interface SearchResults {
  Title: string;
  Year: string;
  Poster: string;
  Type: string;
  imdbID: string;
}

interface MoviesData {
  Search: SearchResults[];
}

export const Movies: FC = () => {
  const [movies, setMovies] = useState<SearchResults[]>([]);
  const [search, setSearch] = useState('');
  const [error, setError] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [page, setPage] = useState(1);
  const bottomRef = useRef<Element | null>(null);
  const moviesCountRef = useRef(0);

  async function getData(searchVal?: string) {
    setIsLoading(true);
    const data: MoviesData = await getMovies(searchVal ?? search, page);

    if (moviesCountRef.current === 0) {
      moviesCountRef.current = movies.length + data.Search.length;
    }

    if (data.Search) {
      setMovies((movieList) => [...movieList, ...data.Search]);
    }
    setIsLoading(false);
  }

  const reset = () => {
    setMovies([]);
    setPage(1);
    moviesCountRef.current = 0;
  };

  const onClearList = () => {
    reset();
  };

  const onMoviesUpdate = (data: SearchResults[], search: string) => {
    setSearch(search);
    moviesCountRef.current = movies.length + data.length;
    setMovies((moviesList) => [...moviesList, ...data]);
    setError('');
  };

  const onMoviesUpdateError = (error: any) => {
    setError(error);
  };

  const handleObserver = (entities: any[]) => {
    const [target] = entities;

    if (target.isIntersecting && moviesCountRef.current > 0) {
      setPage((page) => page + 1);
    }
  };

  const clearStorage = () => {
    sessionStorage.clear();
  };

  useEffect(() => {
    const options = {
      root: null,
      rootMargin: '40px',
      threshold: 0.2,
    };

    const observer = new IntersectionObserver(handleObserver, options);

    if (bottomRef && bottomRef.current) {
      observer.observe(bottomRef.current);
    }

    return () => {
      if (bottomRef.current) {
        observer.unobserve(bottomRef.current);
      }
    };
  }, [bottomRef]);

  useEffect(() => {
    const searchVal = sessionStorage.getItem('search');
    if (searchVal) {
      setSearch(searchVal);
      getData(searchVal);
    }

    window.addEventListener('beforeunload', clearStorage);

    return () => {
      window.removeEventListener('beforeunload', clearStorage);
    };
  }, []);

  useEffect(() => {
    try {
      if (search.length && page > 1) {
        getData();
      }
    } catch (e) {
      console.error('Unable to retrieve Movies', e);
      setError(e.toString());
    }
  }, [page]);

  return (
    <MoviesContainer>
      <SearchBar
        page={page}
        onClearList={onClearList}
        onUpdate={onMoviesUpdate}
        onError={onMoviesUpdateError}
      />
      {!movies.length && !error && (
        <WelcomeText>
          No results found. Please type in the search bar above.
        </WelcomeText>
      )}
      {error && <ErrorText>Error retrieving movies</ErrorText>}
      {movies.map((movie) => {
        return <MovieItem key={movie.imdbID} data={movie} />;
      })}
      {isLoading &&
        [1, 2, 3, 4, 5, 6].map((x) => (
          <MovieItem key={x} data={{}} isPlaceHolder={isLoading} />
        ))}
      <Bottom ref={bottomRef} />
    </MoviesContainer>
  );
};

export default Movies;
