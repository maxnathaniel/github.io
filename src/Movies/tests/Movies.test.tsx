import React, { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { ThemeProvider, theme } from 'src/global';

import { Movies } from '../components/Movies';

/**
 * Utility function that mocks the `IntersectionObserver` API. Necessary for components that rely
 * on it, otherwise the tests will crash. Recommended to execute inside `beforeEach`.
 * @param intersectionObserverMock - Parameter that is sent to the `Object.defineProperty`
 * overwrite method. `jest.fn()` mock functions can be passed here if the goal is to not only
 * mock the intersection observer, but its methods.
 */
export function setupIntersectionObserverMock({
  root = null,
  rootMargin = '',
  thresholds = [],
  disconnect = () => null,
  observe = () => null,
  takeRecords = () => [],
  unobserve = () => null,
} = {}): void {
  class MockIntersectionObserver implements IntersectionObserver {
    readonly root: Element | null = root;
    readonly rootMargin: string = rootMargin;
    readonly thresholds: ReadonlyArray<number> = thresholds;
    disconnect: () => void = disconnect;
    observe: (target: Element) => void = observe;
    takeRecords: () => IntersectionObserverEntry[] = takeRecords;
    unobserve: (target: Element) => void = unobserve;
  }

  Object.defineProperty(window, 'IntersectionObserver', {
    writable: true,
    configurable: true,
    value: MockIntersectionObserver,
  });

  Object.defineProperty(global, 'IntersectionObserver', {
    writable: true,
    configurable: true,
    value: MockIntersectionObserver,
  });
}

beforeEach(() => {
  return new setupIntersectionObserverMock();
});

const history = createMemoryHistory();
const route = '/movie/1551';
history.push(route);

describe('<Movies />', () => {
  it('should render the various inner Elements', () => {
    const { container, getByText } = render(
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <Movies />
        </ThemeProvider>
      </Router>
    );
    expect(
      getByText('No results found. Please type in the search bar above.')
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
