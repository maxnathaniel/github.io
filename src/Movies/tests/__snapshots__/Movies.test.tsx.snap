// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Movies /> should render the various inner Elements 1`] = `
<div>
  <div
    class="sc-gKsewC jRRddb"
  >
    <div
      class="sc-gsTCUz iWCQzw"
    >
      <h1
        class="sc-dlfnbm SSaae"
      >
        MOVIES
      </h1>
      <div
        class="sc-hKgILt fHYTMW"
      >
        <div
          class="ui small fluid icon input"
          data-testid="search-input"
        >
          <input
            placeholder="Input Movie Title..."
            type="text"
            value=""
          />
          <i
            aria-hidden="true"
            class="search icon"
          />
        </div>
      </div>
    </div>
    <p
      class="sc-iBPRYJ gsvDRP"
    >
      No results found. Please type in the search bar above.
    </p>
    
    <div
      class="sc-pFZIQ"
    />
  </div>
</div>
`;
